Add doc describing FIPS 140-3 compliance (#2707)
kyma-project/kyma-environment-broker: docs/contributor/01-06-fips-compliance.md

# FIPS 140-3 Compliance

This document describes the current FIPS 140-3 compliance of the Kyma Environment Broker (KEB).

## Current Status

KEB is compiled and executed in the FIPS 140-3 mode using the Go FIPS module and runtime controls. The build and runtime configuration is defined in the KEB’s [Dockerfile](../../Dockerfile.keb).

### Build-time controls

KEB is built with the Go FIPS module enabled. The Dockerfile has a **build** stage where the **RUN** command includes the **GOFIPS140=v1.0.0** environment variable:

```Dockerfile
RUN --mount=type=cache,target=/root/.cache/go-build \
		--mount=type=cache,target=/go/pkg \
		CGO_ENABLED=0 GOFIPS140=v1.0.0 GOOS=$TARGETOS GOARCH=$TARGETARCH \
		go build -ldflags "-X main.Version=${VERSION}" \
		-o /bin/kyma-env-broker ./cmd/broker/
```

The **GOFIPS140=v1.0.0** environment variable enables building against the Go FIPS 140 module and configures the application to run in the FIPS 140-3 mode.

> ### Runtime controls

Reviewer: ```suggestion
### Runtime Controls
```

> The final image enforces FIPS-only operation via the **GODEBUG** environment variable:

Author: suggestion applied

```Dockerfile
ENV GODEBUG=fips140=only,tlsmlkem=0
```

> The **fips140=only** option restricts cryptographic operations to those approved by the Go FIPS module. The **tlsmlkem=0** option disables the post-quantum key exchange mechanism X25519MLKEM768 to avoid non-FIPS algorithms.

Reviewer: Again, this suggestion is based on the assumption that these are parameter/variable-value pairs. If I'm wrong, please ignore it.

> The **fips140=only** option restricts cryptographic operations to those approved by the Go FIPS module. The **tlsmlkem=0** option disables the post-quantum key exchange mechanism X25519MLKEM768 to avoid non-FIPS algorithms.

Reviewer 2: > The X25519 component of the hybrid X25519MLKEM768 key exchange is not a FIPS-approved algorithm

```suggestion
The **fips140=only** option restricts cryptographic operations to those approved by the Go FIPS module. The **tlsmlkem=0** option disables the post-quantum key exchange mechanism X25519MLKEM768 to avoid usage of not FIPS-approved algorithms.
```

This combination ensures the broker runs exclusively with FIPS-approved algorithms provided by Go’s FIPS module.

## Approved Cryptographic Algorithms

> For the authoritative list and requirements around approved algorithms and testing, refer to [NIST SP 800‑140C supplemental information](https://csrc.nist.gov/projects/cryptographic-module-validation-program/sp-800-140-series-supplemental-information/sp800-140c).

Author: Changed the link title to: **SP 800-140C: Approved Security Functions**

> KEB relies on the Go FIPS module’s set of approved algorithms when **GODEBUG=fips140=only** is active. Any attempt to use non-approved algorithms will be blocked at the runtime.

Reviewer: Once again, I believe the formatting should be changed to code font for a variable-value pair.

> KEB relies on the Go FIPS module’s set of approved algorithms when **GODEBUG=fips140=only** is active. Any attempt to use non-approved algorithms will be blocked at the runtime.

Author: suggestion applied

## PostgreSQL Password Authentication Requirements

> When connecting to PostgreSQL, configure password authentication to use SCRAM-SHA-256 and avoid legacy MD5. The required configuration steps are described in [PostgreSQL documentation](https://www.postgresql.org/docs/current/auth-password.html):

Author: suggestion applied

> When connecting to PostgreSQL, configure password authentication to use SCRAM-SHA-256 and avoid legacy MD5. The required configuration steps are described in [PostgreSQL documentation](https://www.postgresql.org/docs/current/auth-password.html):

Author: suggestion applied

> To upgrade an existing installation from md5 to scram-sha-256, after having ensured that all client libraries in use are new enough to support SCRAM, set password_encryption = 'scram-sha-256' in postgresql.conf, make all users set new passwords, and change the authentication method specifications in pg_hba.conf to scram-sha-256.

> See [PostgreSQL documentation](https://www.postgresql.org/docs/current/auth-password.html) for details.

Reviewer: ```suggestion
For more information, see [PostgreSQL documentation](https://www.postgresql.org/docs/current/auth-password.html).
```
